howso.amlg

```diff
@@ -90,18 +92,130 @@
 			(retrieve_from_entity trainee "pointVersion")
 		)
 
+	;helper method used to determine the full hierarchy path to a subtrainee used to create or load subtrainees
+	;outputs path in the format of: (list trainee ".trainee_container" child ".trainee_container" grandchild ...)
+	#!determine_child_trainee_path
+	(if child_trainee
+		(seq
+			(if (not (contains_entity trainee))
+				(conclude "Specified invalid trainee.")
+			)
+
+			(if (!= (null) (call_entity trainee "GetTraineePath" (assoc trainee child_trainee)))
+				(conclude "Specified child_trainee is already in the hierarchy.")
+			)
+
+			(if parent_trainee
+				(let
+					(assoc
+						parent_trainee_path (first (call_entity trainee "GetTraineePath" (assoc trainee parent_trainee)))
+					)
+					(if (= (null) parent_trainee_path)
+						(conclude "Specified invalid parent_trainee.")
+					)
+
+					;set path to the child through whatever the path is to its parent
+					(append trainee parent_trainee_path ".trainee_container" child_trainee)
+				)
+
+				;else it's a direct child of this trainee
+				(list trainee ".trainee_container" child_trainee)
+			)
+		)
+	)
+
+	;Pull the hierarchy for a trainee, returns an assoc of:
+	;'hierarchy_map' which reperesents the hierarchy tree and 'is_contained_map' which is an assoc of subtrainee -> is_contained flag
+	;parameters:
+	; trainee: id of trainee whose hierarchy info to return
+	#get_hierarchy
+		(call !return (assoc
+			payload
+				(assoc
+					"hierarchy_map" (retrieve_from_entity trainee "hierarchyMap")
+					"is_contained_map" (retrieve_from_entity trainee "childTraineeIsContainedMap")
+				)
+		))
+
+	;Execute any method in the API directly on any child trainee of trainee
+	;parameters:
+	; trainee: id of trainee
+	; child_trainee: id of child trainee to execute method
+	; method: name of method to execute
+	; payload: payload to pass to the method
+	#execute_method
+	(declare
+		(assoc payload (assoc))
+
+		(if (= (null) trainee)
+			(conclude (call !return (assoc errors (list "Invalid trainee specified."))))
```

Comment:
This error should probably say "No trainee specified.", (null) is certainly invalid, but I feel this error would be misleading